src/defrag.c

```diff
@@ -1497,7 +1513,20 @@ static void beginDefragCycle(void) {
     addDefragStage(defragStagePubsubKvstore, defrag_pubsubshard_ctx);
 
     addDefragStage(defragLuaScripts, NULL);
-    addDefragStage(defragModuleGlobals, NULL);
+
+    /* Add stages for modules. */
+    dictIterator *di = dictGetIterator(modules);
+    dictEntry *de;
+    while ((de = dictNext(di)) != NULL) {
+        struct RedisModule *module = dictGetVal(de);
+        if (module->defrag_cb) {
+            defragModuleCtx *ctx = zmalloc(sizeof(defragModuleCtx));
+            ctx->module = module;
+            ctx->cursor = 0;
+            addDefragStage(defragModuleGlobals, ctx);
```

Comment:
so we add a defrag stage per module, right?
and we re-create the stages on each sweep we start?

what happens if a module is un-loaded during a sweep?

i wonder if we have any observability feature to know when one is slow, or hang, and which one is it.
probably not mandatory.

Reply:
> so we add a defrag stage per module, right?

yes

> and we re-create the stages on each sweep we start?

we only recreate all the stages when start a new defragmentaion.

> what happens if a module is un-loaded during a sweep?

good catch, we can't store the pointer of the module in the context, instead we need find a way to get the module by id or by name.

> i wonder if we have any observability feature to know when one is slow, or hang, and which one is it.
probably not mandatory.

not now, I think in the future we can count the elapsed time of each stage.